lib/remote/objectqueryhandler.cpp

```diff
@@ -262,24 +265,21 @@ bool ObjectQueryHandler::HandleRequest(
 				continue;
 
 			Type::Ptr reflectionType = joinedObj->GetReflectionType();
-			Expression::Ptr permissionFilter;
-
 			auto it = typePermissions.find(reflectionType.get());
 			bool granted;
 
 			if (it == typePermissions.end()) {
 				String permission = "objects/query/" + reflectionType->GetName();
 
-				Expression *filter = nullptr;
-				granted = FilterUtility::HasPermission(user, permission, &filter);
-				permissionFilter = filter;
+				std::unique_ptr<Expression> permissionFilter;
+				granted = FilterUtility::HasPermission(user, permission, &permissionFilter);
 
-				typePermissions.insert({reflectionType.get(), std::make_pair(granted, permissionFilter)});
+				permissionFilters = typePermissions.insert({reflectionType.get(), std::make_pair(granted, std::move(permissionFilter))}).first;
```

Comment:
You can just assign the returned iterator to `it`, drop the else branch and do this after the if:

```c++
bool granted = it->second.first;
const unique_ptr<Expression>& permissionFilter = it->second.second;
```

So no need for the typedef and second iterator variable above.

By the way, [C++17 magic](https://en.cppreference.com/w/cpp/language/structured_binding) would allow this: `const auto& [granted, permissionFilter] = it->seconds;`

Reply:
Except that we might want to backport this as it should fix a memory leak in `FilterUtility::GetFilterTargets()`, so we probably need a C++11 version anyways.